controllers/node_controller_test.go

```diff
@@ -0,0 +1,178 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		ctx, cancel := context.WithCancel(ctx)
+		stopFunc = cancel
+		go func() {
+			errCh <- mgr.Start(ctx)
+		}()
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	AfterEach(func() {
+		stopFunc()
+		Expect(<-errCh).NotTo(HaveOccurred())
+	})
+
+	setupResources := func(ctx context.Context, suffix string) (
+		corev1.Node, corev1.PersistentVolumeClaim, topolvmv1.LogicalVolume) {
+		node := corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node" + suffix,
+				Finalizers: []string{
+					topolvm.GetNodeFinalizer(),
+				},
+			},
+		}
+		err := k8sClient.Create(ctx, &node)
+		Expect(err).NotTo(HaveOccurred())
+
+		sc := storegev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "sc" + suffix,
+			},
+			Provisioner: topolvm.GetPluginName(),
+		}
+		err = k8sClient.Create(ctx, &sc)
+		Expect(err).NotTo(HaveOccurred())
+
+		ns := createNamespace()
+		pvc := corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pvc" + suffix,
+				Namespace: ns,
+				Annotations: map[string]string{
+					AnnSelectedNode: node.Name,
+				},
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				StorageClassName: &sc.Name,
+				AccessModes: []corev1.PersistentVolumeAccessMode{
+					corev1.ReadWriteOnce,
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: *resource.NewQuantity(1, resource.BinarySI),
+					},
+				},
+			},
+		}
+		err = k8sClient.Create(ctx, &pvc)
+		Expect(err).NotTo(HaveOccurred())
+
+		lv := topolvmv1.LogicalVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "lv" + suffix,
+			},
+			Spec: topolvmv1.LogicalVolumeSpec{
+				NodeName: node.Name,
+			},
+		}
+		err = k8sClient.Create(ctx, &lv)
+		Expect(err).NotTo(HaveOccurred())
+
+		return node, pvc, lv
+	}
+
+	It("should delete PVC and LogicalVolume when the node is deleted if the finalizer is not skipped", func() {
+		startReconciler(false)
+
+		ctx := context.Background()
+
+		// Setup
+		node, pvc, lv := setupResources(ctx, "-do-finalizer")
+
+		// Exercise
+		err := k8sClient.Delete(ctx, &node)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Verify
+		Eventually(func(g Gomega) error {
+			err := k8sClient.Get(ctx, client.ObjectKeyFromObject(&node), &node)
+			if err == nil || !apierrors.IsNotFound(err) {
+				return errors.New("Node is not deleted")
+			}
+
+			err = k8sClient.Get(ctx, client.ObjectKeyFromObject(&pvc), &pvc)
+			if err == nil {
+				// k8s may add the pvc-protection finalizer, so check DeletionTimestamp.
+				if pvc.DeletionTimestamp == nil {
+					return fmt.Errorf("delete API is not called")
+				}
+			} else if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("PVC is not deleted: %w", err)
+			}
+
+			err = k8sClient.Get(ctx, client.ObjectKeyFromObject(&lv), &lv)
+			if err == nil || !apierrors.IsNotFound(err) {
+				return errors.New("LV is not deleted")
+			}
+
+			return nil
+		}).ShouldNot(HaveOccurred())
```

Comment:
```suggestion
		}).Should(Succeed())
```

may be more readable as a sentence. There is one more point that accepts this improvement.

Reply:
I'll fix it.

Reply:
https://github.com/topolvm/topolvm/compare/a5bb8ccf83bccfe41bc38be95f768b0667044633..2ac772badcc965ea370dfd429e5c8dd2db546453